src/Articulate/Components/ContentPublishedHandler.cs

```diff
@@ -0,0 +1,55 @@
+using Umbraco.Cms.Core.Events;
+using Umbraco.Cms.Core.Notifications;
+using Umbraco.Extensions;
+using Articulate.Routing;
+using Umbraco.Cms.Core.Web;
+using Umbraco.Cms.Core;
+using Microsoft.AspNetCore.Http;
+using J2N.Collections.Generic;
+
+namespace Articulate.Components
+{
+
+    public class ContentPublishedHandler : INotificationHandler<ContentPublishedNotification>
+    {
+        private readonly IUmbracoContextFactory _umbracoContextFactory;
+        private readonly IHttpContextAccessor _httpContextAccessor;
+        private readonly ArticulateRouter _articulateRouter;
+
+        public ContentPublishedHandler(IUmbracoContextFactory umbracoContextFactory, IHttpContextAccessor httpContextAccessor, ArticulateRouter articulateRouter)
+        {
+            _umbracoContextFactory = umbracoContextFactory;
+            _httpContextAccessor = httpContextAccessor;
+            _articulateRouter = articulateRouter;
+        }
+
+        public void Handle(ContentPublishedNotification notification)
+        {
+            var e = notification;
+
+            foreach (var c in e.PublishedEntities)
+            {
+                if (!c.ContentType.Alias.InvariantEquals(ArticulateConstants.ArticulateContentTypeAlias))
+                    continue;
+
+                // Using WereDirty as the content has been published/saved now
+                var dirtyProps = c.GetWereDirtyProperties();
+                var urlPropsToCheck = new List<string>{ "categoriesUrlName", "tagsUrlName", "searchUrlName" };
+                if (dirtyProps.ContainsAny(urlPropsToCheck))
+                {
+                    var httpCtx = _httpContextAccessor.GetRequiredHttpContext();
```

Comment:
If this doesn't error, then we could assume that this notification always fires within the context of a web request. However, perhaps there's not a guarantee of this - @bergmania would you happen to know this off the top of your head? Else the only other way to do this would be as a middleware and this handler would set some sort of flag that would tell the middleware that it needs to re route everything.

Reply:
ContentPublishedNotification can happen in a background thread when `ScheduledPublishing` is executing..

Reply:
Little naive me @bergmania does that mean that using `ar httpCtx = _httpContextAccessor.GetRequiredHttpContext();` is the right approach then?

If so then would you still be happy with this @Shazwazza ?

Reply:
HttpContext cannot be used in background jobs, as there is no request, when these are running.

Reply:
@warrenbuckley so we cannot do it this way and we'll have to create a middleware that does this. This handler will need to set a flag (can be in memory/cache) to say: "we need to rebuild the articulate routes on next request". Then on each request the middleware checks for this flag and if it's there, we do this logic to rebuild the articulate routes and clear the flag.

Reply:
Arghh dammnit @bergmania - shame as this was working nicely for me locally.
Last couple of weeks has been busy (priorities changed), will try to get the two pending PRs with feedback done soon though